### PR TITLE
Feature/block integration test

### LINKED ISSUE #1567: E2E Testing - Block Model

### PR DESCRIPTION
**Pull request by HarshitJoshi**
add E2E tests for block model and update the test blocks

Fixes #1567

### CODE REVIEW

File: packages/bitcore-node/test/integration/models/block.integration.ts [hunk 1]
```diff
@@ -0,0 +1,611 @@
...
+
+      const result = await BlockModel.getLocalTip({ chain: 'ETH', network: 'regtest' });
+
+      expect(result.height).to.equal(0);
+
+    });
+  });
+
+  describe('getLocatorHashes', () => {
+    it('should return the highest processed block hash for the last thirty blocks for a particular chain and network', async () => {
```
Comment (DanielRHayes) [inline packages/bitcore-node/test/integration/models/block.integration.ts]: If you test title says that this should be for the last 30 blocks then you need to insert 30 blocks. For 30 it's ok to use a simple loop.  It's also ok if you don't have a real hash for this test and just do `hash1` `hash2` and so on.

```
it('should return the highest....', async(){

for(let i = 1; i < 31; i++){
await BlockModel.create({processed: true, chain: 'BTC', network: 'regtest', height: i , hash: `hash${i}`);
}

// then your test and expects

}

```

File: packages/bitcore-node/test/integration/models/block.integration.ts [hunk 2]
```diff
@@ -0,0 +1,611 @@
...
+
+      expect(result).to.deep.equal([
+        '3577099967d5036c116dcd7810d637334a011523c9f20a4092018feacc3b8837',
+        '74b8e5f1231638c9ba880d2252ad451e6b675a7afde815f2702dd78c95152769',
+        '3d7c133d28353247312c9673a8d60151c0858486fa87fc16764f5e282f06f9f7',
+        '1e980bdd683513b2cdeaf81985f1f52e17175d3ce34e3be56e0c210eed0a21a3'
+      ]);
+
+    });
+    it('should return block hash with sixty four zeros if processed block count less than two is found for a particular chain and network', async () => {
```
Comment (DanielRHayes) [inline packages/bitcore-node/test/integration/models/block.integration.ts]: I would use the number `64` instead of typing out `sixty four`, it will be easier to search for.

File: packages/bitcore-node/test/integration/models/block.integration.ts [hunk 3]
```diff
@@ -0,0 +1,611 @@
...
+
+      const result = await BlockModel.getLocatorHashes({ chain: 'BTC', network: 'regtest' });
+
+      expect(result).to.deep.equal([Array(65).join('0')]);
+
+    });
+  });
+
+  describe('handleReorg', () => {
+    it('should return if incoming block hash references previous block hash (no conflict)', async () => {
```
Comment (DanielRHayes) [inline packages/bitcore-node/test/integration/models/block.integration.ts]: maybe `it('should not reorg if the incoming block's prevHash matches the block hash of the current highest block')`

File: packages/bitcore-node/test/integration/models/block.integration.ts [hunk 4]
```diff
@@ -0,0 +1,611 @@
...
+        mintIndex: 0,
+        mintHeight: 5,
+        coinbase: true,
+        value: 500.0,
+        address: 'mkjB6LmjiNfJWgH4aP4v1GkFjRcQTfDSfj',
+        spentTxid: 'eec8570a0c960b19fa6c86c71a06ebda379b86b5fe0be0e64ba83b2e0a3d05a3',
+        spentHeight: 9
+      });
+
+      const localTip = await BlockModel.getLocalTip({ chain: 'BTC', network: 'regtest' });
```
Comment (DanielRHayes) [inline packages/bitcore-node/test/integration/models/block.integration.ts]: remove the expects before `handleReorg is called` just tests `handleReorg`

File: packages/bitcore-node/test/integration/models/block.integration.ts [hunk 5]
```diff
@@ -0,0 +1,611 @@
...
+      const blocks = await BlockModel.find({
+        chain: 'BTC',
+        network: 'regtest'
+      }).exec();
+      expect(blocks.length).to.equal(2);
+
+      const removedBlock = await BlockModel.find({
+        chain: 'BTC',
+        network: 'regtest',
+        hash: '3279069d22ce5af68ef38332d5b40e79e1964b154d466e7fa233015a34c27312',
```
Comment (DanielRHayes) [inline packages/bitcore-node/test/integration/models/block.integration.ts]: remove hash and change height to be what we expect localTip height to be (hard code what we expect it to be)

File: packages/bitcore-node/test/integration/models/block.integration.ts [hunk 6]
```diff
@@ -0,0 +1,611 @@
...
+        chain: 'BTC',
+        network: 'regtest'
+      }).exec();
+      expect(transaction.length).to.equal(1);
+
+      const removedTransaction = await TransactionModel.find({
+        chain: 'BTC',
+        network: 'regtest',
+        blockHeight: {
+          $gte: localTip.height
```
Comment (DanielRHayes) [inline packages/bitcore-node/test/integration/models/block.integration.ts]: hard code instead of localTip

File: packages/bitcore-node/test/integration/models/block.integration.ts [hunk 7]
```diff
@@ -0,0 +1,611 @@
...
+        }
+      }).exec();
+      expect(removedTransaction).to.be.an('array').that.is.empty;
```
Comment (DanielRHayes) [inline packages/bitcore-node/test/integration/models/block.integration.ts]: `expect(removedTransaction.length).to.equal(0)`

File: packages/bitcore-node/test/integration/models/block.integration.ts [hunk 8]
```diff
@@ -0,0 +1,611 @@
...
+        chain: 'BTC',
+        network: 'regtest',
+      }).exec();
+      expect(coinModel.length).to.equal(1);
+
+      const removedCoin = await CoinModel.find({
+        chain: 'BTC',
+        network: 'regtest',
+        mintHeight: {
+          $gte: localTip.height
```
Comment (DanielRHayes) [inline packages/bitcore-node/test/integration/models/block.integration.ts]: hard code instead of localTip

File: packages/bitcore-node/test/integration/models/block.integration.ts [hunk 9]
```diff
@@ -0,0 +1,611 @@
...
+        }
+      });
+      expect(removedCoin).to.be.an('array').that.is.empty;
```
Comment (DanielRHayes) [inline packages/bitcore-node/test/integration/models/block.integration.ts]: `expect(removedTransaction.length).to.equal(0)`

File: packages/bitcore-node/test/integration/models/block.integration.ts [hunk 10]
```diff
@@ -0,0 +1,611 @@
...
+      expect(transaction[0].chain).to.equal('BTC');
+      expect(transaction[0].network).to.equal('regtest');
+      expect(transaction[0].txid).to.equal('08e23107e8449f02568d37d37aa76e840e55bbb5f100ed8ad257af303db88c08');
+      expect(transaction[0].blockHash).to.equal('64bfb3eda276ae4ae5b64d9e36c9c0b629bc767fb7ae66f9d55d2c5c8103a929');
+      expect(transaction[0].blockHeight).to.equal(1);
+      expect(transaction[0].coinbase).to.equal(true);
+      expect(transaction[0].locktime).to.equal(0);
+      expect(transaction[0].size).to.equal(0);
+      // TODO: assertion for block times
+      expect(transaction[0].wallets).to.be.an('array').that.is.empty;
```
Comment (DanielRHayes) [inline packages/bitcore-node/test/integration/models/block.integration.ts]: `expect(length of wallets).to.equal(0)`

File: packages/bitcore-node/test/integration/models/block.integration.ts [hunk 11]
```diff
@@ -0,0 +1,611 @@
...
+      expect(transaction[0].chain).to.equal('BTC');
+      expect(transaction[0].network).to.equal('regtest');
+      expect(transaction[0].txid).to.equal('08e23107e8449f02568d37d37aa76e840e55bbb5f100ed8ad257af303db88c08');
+      expect(transaction[0].blockHash).to.equal('64bfb3eda276ae4ae5b64d9e36c9c0b629bc767fb7ae66f9d55d2c5c8103a929');
+      expect(transaction[0].blockHeight).to.equal(9);
+      expect(transaction[0].coinbase).to.equal(true);
+      expect(transaction[0].locktime).to.equal(0);
+      expect(transaction[0].size).to.equal(0);
+      // TODO: assertion for block times
+      expect(transaction[0].wallets).to.be.an('array').that.is.empty;
```
Comment (DanielRHayes) [inline packages/bitcore-node/test/integration/models/block.integration.ts]: expect length of wallets to be 0

File: packages/bitcore-node/test/integration/models/block.integration.ts [hunk 12]
```diff
@@ -0,0 +1,611 @@
...
+        },
+        chain: 'BTC',
+        network: 'regtest'
+      });
+
+      const result = await BlockModel.find({ chain: 'BTC', network: 'regtest'}).exec();
+      expect(result.length).to.equal(3);
+
+    });
+    it('should return if localTip height is zero', async () => {
```
Comment (DanielRHayes) [inline packages/bitcore-node/test/integration/models/block.integration.ts]: `it('should not reorg if localTip height is zero')`

File: packages/bitcore-node/test/integration/models/block.integration.ts [hunk 13]
```diff
@@ -0,0 +1,611 @@
...
+          merkleRoot: '8c29860888b915715878b21ce14707a17b43f6c51dfb62a1e736e35bc5d8093f',
+          bits: parseInt('207fffff', 16).toString(),
+          nonce: '3'
+        },
+        chain: 'ETH',
+        network: 'regtest'
+      });
+
+      const result = await BlockModel.find({ chain: 'BTC', network: 'regtest'}).exec();
+      expect(result.length).to.equal(3);
```
Comment (DanielRHayes) [inline packages/bitcore-node/test/integration/models/block.integration.ts]: let's not add any blocks on this test and then call `handleReorg` and expect the length to be zero

File: packages/bitcore-node/test/integration/models/block.integration.ts [hunk 14]
```diff
@@ -0,0 +1,611 @@
...
+      const removedCoin = await CoinModel.find({
+        chain: 'BTC',
+        network: 'regtest',
+        mintHeight: {
+          $gte: localTip.height
+        }
+      });
+      expect(removedCoin).to.be.an('array').that.is.empty;
+
+      const updateCoin = await CoinModel.find({
```
Comment (DanielRHayes) [inline packages/bitcore-node/test/integration/models/block.integration.ts]: maybe call this `unspentCoins` or something similar

File: packages/bitcore-node/test/integration/models/block.integration.ts [hunk 15]
```diff
@@ -0,0 +1,611 @@
...
+        },
+        chain: 'ETH',
+        network: 'regtest'
+      });
+
+      const result = await BlockModel.find({ chain: 'BTC', network: 'regtest'}).exec();
+      expect(result.length).to.equal(3);
+
+    });
+    it('should successfully handle reorg', async () => {
```
Comment (DanielRHayes) [inline packages/bitcore-node/test/integration/models/block.integration.ts]: we should add comments to this test to explain what we are doing and checking in the various parts

File: packages/bitcore-node/test/data/test-block.ts [hunk 16]
```diff
@@ -36,3 +36,111 @@ export const TEST_BLOCK: Bitcoin.Block = {
...
+          '0000000000000000000000000000000000000000000000000000000000000a98',
+        prevHash:
+          '64bfb3eda276ae4ae5b64d9e36c9c0b629bc767fb7ae66f9d55d2c5c8103a929'
+      };
+    }
+  }
+};
+export const TEST_BLOCK_2: Bitcoin.Block = {
+  hash: '2a883ff89c7d6e9302bb4a4634cd580319a4fd59d69e979b344972b0ba042b86',
+  transactions: [TEST_TX_3],
```
[the remark this answers is not in the document]
Comment (HarshitJoshi) [reply packages/bitcore-node/test/data/test-block.ts]: resolved

File: packages/bitcore-node/test/integration/models/block.integration.ts [hunk 17]
```diff
@@ -0,0 +1,611 @@
...
+import { BlockModel } from '../../../src/models/block';
+import { TransactionModel } from '../../../src/models/transaction';
+import { CoinModel } from '../../../src/models/coin';
+import { TEST_BLOCK } from '../../data/test-block';
+import logger from '../../../src/logger';
+
+describe('Block Model', function () {
+
+  beforeEach(async () => {
+    await resetModel(BlockModel);
```
[the remark this answers is not in the document]
Comment (HarshitJoshi) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: resolved

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 1 above)
Comment (HarshitJoshi) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: resolved

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 2 above)
Comment (HarshitJoshi) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: updated

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 12 above)
Comment (HarshitJoshi) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: resolved

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 13 above)
Comment (HarshitJoshi) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: resolved

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 14 above)
Comment (HarshitJoshi) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: resolved

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 7 above)
Comment (HarshitJoshi) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: updated

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 9 above)
Comment (HarshitJoshi) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: updated

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 10 above)
Comment (HarshitJoshi) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: updated

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 11 above)
Comment (HarshitJoshi) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: updated

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 4 above)
Comment (HarshitJoshi) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: resolved

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 5 above)
Comment (HarshitJoshi) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: resolved

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 6 above)
Comment (HarshitJoshi) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: resolved

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 8 above)
Comment (HarshitJoshi) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: resolved

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 3 above)
Comment (HarshitJoshi) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: updated

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 15 above)
Comment (HarshitJoshi) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: updated

File: packages/bitcore-node/test/data/test-block.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (DanielRHayes) [reply packages/bitcore-node/test/data/test-block.ts]: confirmed

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (DanielRHayes) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: confirmed

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 10 above)
Comment (DanielRHayes) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: confirmed

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 11 above)
Comment (DanielRHayes) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: confirmed

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 1 above)
Comment (DanielRHayes) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: confirmed

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 2 above)
Comment (DanielRHayes) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: confirmed

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 3 above)
Comment (DanielRHayes) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: confirmed

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 12 above)
Comment (DanielRHayes) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: confirmed

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 15 above)
Comment (DanielRHayes) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: confirmed

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 4 above)
Comment (DanielRHayes) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: confirmed

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 6 above)
Comment (DanielRHayes) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: confirmed

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 7 above)
Comment (DanielRHayes) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: confirmed

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 9 above)
Comment (DanielRHayes) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: confirmed

File: packages/bitcore-node/test/integration/models/block.integration.ts (hunk 14 above)
Comment (DanielRHayes) [reply packages/bitcore-node/test/integration/models/block.integration.ts]: confirmed
